lib/protobuf/message.rb

```diff
@@ -80,37 +70,26 @@ def each_field
     end
 
     def each_field_for_serialization
-      self.class.all_fields.each do |field|
-        value = @values[field.fully_qualified_name]
-        if value.nil?
-          fail ::Protobuf::SerializationError, "Required field #{self.class.name}##{field.name} does not have a value." if field.required?
-          next
-        end
-        if field.map?
-          # on-the-wire, maps are represented like an array of entries where
-          # each entry is a message of two fields, key and value.
-          array = Array.new(value.size)
-          i = 0
-          value.each do |k, v|
-            array[i] = field.type_class.new(:key => k, :value => v)
-            i += 1
-          end
-          value = array
-        end
+      _protobuf_message_required_field_tags.each do |tag|
+        fail ::Protobuf::SerializationError, "Required field #{self.class.name}##{_protobuf_message_field[tag].name} does not have a value."
```

Comment:
Should this have a conditional check before raising an error?

Reply:
what conditional check would you recommend? only the required field tags are in the array so if anything is present in the array then it's a problem

Reply:
I see now. `_protobuf_message_required_field_tags` only stores the unset field tags. I think it would be easier to understand if the function mentioned it only contains unset attributes.